woocommerce/woocommerce
# Prevent "woocommerce_order_status_X' hooks to be called when Order save() action is cancelled

### Changes proposed in this Pull Request:

Would close #46886.

### How to test the changes in this Pull Request:
1) Navigate to `/wp-admin/admin.php?page=wc-orders&action=new` and hit "create" button to create an order
2) in `functions.php`, add an action hook to ensure that no order has reach the "refunded" status
```
add_action(
    hook_name: 'woocommerce_order_status_refunded',
    callback: static function ($id) {
        die('If you see this, this fix doesn\'t work :-)');
    }
);
```
3) in functions.php, add an action hook on the `woocommerce_before_order_object_save` to simulate a check that prevent any order to be saved with the "refunded" status
```
add_action(
    hook_name: 'woocommerce_before_order_object_save',
    callback: static function (WC_Order $order) {
        $changes = $order->get_changes();
        if (!empty($changes) && isset($changes['status']) && $changes['status'] === 'refunded') {
            throw new \RuntimeException('to prove my point');
        }
    }
);
```
4) in the backoffice order page, change ist status to "refunded" in the select list and hit the "update" button.
5) page should refresh and still show the previous order status (= NOT refunded)

----

Another way to do actions 1), 4) & 5) would be to use the following code
```
$order = new Wc_Order();
$order->set_status('draft');
$order->save();
$order->set_status('refunded');
$order->save();
```
